### PR TITLE
[FLINK-14983][table-common] Add interfaces for input type inference

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This adds a similar class as Calcite's `SqlOperandTypeInference` to Flink's type inference.

For FLIP-65, we will need to implement those interfaces as the planner needs infer the `DataType`s out of logical types that come from the logical query. This is also beneficial to support the NULL literal in the future.

## Brief change log

- Introduction of `InputTypeStrategy`
- Update of the `TypeInferenceUtil`
- Integration into `ResolveCallByArgumentsRule` (not used yet)
- Implementation of built-in `InputTypeStrategies`

## Verifying this change

This change added tests and can be verified as follows: `InputTypeStrategiesTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+			// incomplete inference
+			TestSpec
+				.forInputStrategy(InputTypeStrategies.NOP)
+				.actualArgumentTypes(DataTypes.NULL(), DataTypes.STRING(), DataTypes.NULL())
+				.expectErrorMessage("Invalid use of untyped NULL in arguments."),
+
+			// typed arguments help inferring a type
+			TestSpec
+				.forInputStrategy(InputTypeStrategies.NOP)
+				.typedArguments(DataTypes.INT(), DataTypes.STRING(), DataTypes.BOOLEAN())
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: Could we combine this somehow with the input type strategy? Either accept a full `TypeInference` in `forTypeInference` or add method `forInputStrategy(InputTypeStrategy, DataType...)`, `forInputStrategy(InputTypeStrategy, SurroundingInfo)`

I spent some time understanding its relation with `actualArgumentTypes` and `expectArgumentTypes`. It actually describes the inputStrategy, no?

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+				.actualArgumentTypes(DataTypes.NULL(), DataTypes.STRING(), DataTypes.NULL())
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: `actualArgumentTypes` -> `callArgumentTypes`? `callContext`? `calledWithTypes`?

At first I thought this somehow describes an assertion.

Comment (twalthr) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: I chose `calledWithArgumentTypes`.

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: I'm not sure if this improves the readability of the tests. In theory we could also have all three `InputTypeStrategy`, `SurroundingInfo` and `typedArguments`. The latter two just influence the behavior of the input type strategy. How about we rename those methods to `addSurroundingInfo/addTypedArguments` or `withXXX`?

Comment (dawidwys) [reply flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java]: I'm fine with leaving it as it is. I think my confusion was mostly caused by the `actualArgumentTypes`.

**Comment by twalthr** [comment]
Thanks for the review @dawidwys. Merging...

**Pull request closed without merging**